File: src/main/java/no/nav/syfo/exception/ControllerExceptionHandler.java
package no.nav.syfo.exception;

import lombok.extern.slf4j.Slf4j;
import no.nav.security.spring.oidc.validation.interceptor.OIDCUnauthorizedException;
import no.nav.syfo.metric.Metrikk;
import no.nav.syfo.services.exceptions.FastlegeIkkeFunnet;
import no.nav.syfo.services.exceptions.HarIkkeTilgang;
import org.springframework.http.*;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.util.WebUtils;

import javax.inject.Inject;
import javax.validation.ConstraintViolationException;
import javax.ws.rs.ForbiddenException;

@Slf4j
@ControllerAdvice
public class ControllerExceptionHandler {

    private final static String BAD_REQUEST_MSG = "Vi kunne ikke tolke inndataene";
    public final static String FORBIDDEN_MSG = "Har ikke tilgang";
    private final static String INTERNAL_MSG = "Det skjedde en uventet feil";
    private final static String UNAUTHORIZED_MSG = "Autorisasjonsfeil";

    private Metrikk metrikk;

    @Inject
    public ControllerExceptionHandler(Metrikk metrikk) {
        this.metrikk = metrikk;
    }

    @ExceptionHandler({
            Exception.class,
            ConstraintViolationException.class,
            HarIkkeTilgang.class,
            FastlegeIkkeFunnet.class,
            ForbiddenException.class,
            IllegalArgumentException.class,
            OIDCUnauthorizedException.class,
    })
    public final ResponseEntity<ApiError> handleException(Exception ex, WebRequest request) {
        HttpHeaders headers = new HttpHeaders();

        if (ex instanceof OIDCUnauthorizedException) {
            OIDCUnauthorizedException notAuthorizedException = (OIDCUnauthorizedException) ex;

            return handleOIDCUnauthorizedException(notAuthorizedException, headers, request);
        } else if (ex instanceof ForbiddenException) {
            ForbiddenException forbiddenException = (ForbiddenException) ex;

            return handleForbiddenException(forbiddenException, headers, request);
        } else if (ex instanceof IllegalArgumentException) {
            IllegalArgumentException illegalArgumentException = (IllegalArgumentException) ex;

            return handleIllegalArgumentException(illegalArgumentException, headers, request);
        } else if (ex instanceof ConstraintViolationException) {
            ConstraintViolationException constraintViolationException = (ConstraintViolationException) ex;

            return handleConstraintViolationException(constraintViolationException, headers, request);
        } else if (ex instanceof FastlegeIkkeFunnet) {
            FastlegeIkkeFunnet notFoundException = (FastlegeIkkeFunnet) ex;

            return handleFastlegeIkkeFunnetException(notFoundException, headers, request);
        } else {
            HttpStatus status = HttpStatus.INTERNAL_SERVER_ERROR;

            log.warn("Fikk RuntimeException i én av controllerene");
            return handleExceptionInternal(ex, new ApiError(status.value(), INTERNAL_MSG), headers, status, request);
        }
    }

    private ResponseEntity<ApiError> handleConstraintViolationException(ConstraintViolationException ex, HttpHeaders headers, WebRequest request) {
        return handleExceptionInternal(ex, new ApiError(HttpStatus.BAD_REQUEST.value(), BAD_REQUEST_MSG), headers, HttpStatus.BAD_REQUEST, request);
    }

    private ResponseEntity<ApiError> handleForbiddenException(ForbiddenException ex, HttpHeaders headers, WebRequest request) {
        HttpStatus status = HttpStatus.FORBIDDEN;
        return handleExceptionInternal(ex, new ApiError(status.value(), FORBIDDEN_MSG), headers, status, request);
    }

    private ResponseEntity<ApiError> handleIllegalArgumentException(IllegalArgumentException ex, HttpHeaders headers, WebRequest request) {
        HttpStatus status = HttpStatus.BAD_REQUEST;
        return handleExceptionInternal(ex, new ApiError(status.value(), BAD_REQUEST_MSG), headers, status, request);
    }

    private ResponseEntity<ApiError> handleOIDCUnauthorizedException(OIDCUnauthorizedException ex, HttpHeaders headers, WebRequest request) {
        HttpStatus status = HttpStatus.UNAUTHORIZED;
        return handleExceptionInternal(ex, new ApiError(status.value(), UNAUTHORIZED_MSG), headers, status, request);
    }

    private ResponseEntity<ApiError> handleFastlegeIkkeFunnetException(FastlegeIkkeFunnet ex, HttpHeaders headers, WebRequest request) {
        HttpStatus status = HttpStatus.NOT_FOUND;
        return handleExceptionInternal(ex, new ApiError(status.value(), ex.getMessage()), headers, status, request);
    }

    private ResponseEntity<ApiError> handleExceptionInternal(Exception ex, ApiError body, HttpHeaders headers, HttpStatus status, WebRequest request) {
        metrikk.tellHttpKall(status.value());

        if (HttpStatus.INTERNAL_SERVER_ERROR.equals(status)) {
            request.setAttribute(WebUtils.ERROR_EXCEPTION_ATTRIBUTE, ex, WebRequest.SCOPE_REQUEST);
        }

        return new ResponseEntity<>(body, headers, status);
    }
}

File: src/main/kotlin/no/nav/syfo/api/exception/ControllerExceptionHandler.kt
package no.nav.syfo.api.exception

import no.nav.security.token.support.spring.validation.interceptor.JwtTokenUnauthorizedException
import no.nav.syfo.dialogmelding.exception.InnsendingFeiletException
import no.nav.syfo.dialogmelding.exception.PartnerinformasjonIkkeFunnet
import no.nav.syfo.fastlege.expection.*
import no.nav.syfo.metric.Metrikk
import org.slf4j.LoggerFactory
import org.springframework.http.*
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.context.request.WebRequest
import org.springframework.web.util.WebUtils
import javax.inject.Inject
import javax.validation.ConstraintViolationException
import javax.ws.rs.ForbiddenException

@ControllerAdvice
class ControllerExceptionHandler @Inject constructor(
    private val metrikk: Metrikk
) {
    @ExceptionHandler(
        Exception::class,
        ConstraintViolationException::class,
        HarIkkeTilgang::class,
        FastlegeIkkeFunnet::class,
        ForbiddenException::class,
        IllegalArgumentException::class,
        JwtTokenUnauthorizedException::class,
        PartnerinformasjonIkkeFunnet::class,
        InnsendingFeiletException::class
    )
    fun handleException(ex: Exception, request: WebRequest): ResponseEntity<ApiError> {
        val headers = HttpHeaders()
        return when (ex) {
            is JwtTokenUnauthorizedException -> {
                handleJwtTokenUnauthorizedException(ex, headers, request)
            }
            is ForbiddenException -> {
                handleForbiddenException(ex, headers, request)
            }
            is IllegalArgumentException -> {
                handleIllegalArgumentException(ex, headers, request)
            }
            is ConstraintViolationException -> {
                handleConstraintViolationException(ex, headers, request)
            }
            is FastlegeIkkeFunnet -> {
                handleFastlegeIkkeFunnetException(ex, headers, request)
            }
            is PartnerinformasjonIkkeFunnet -> {
                handlePartnerinformasjonIkkeFunnetException(ex, headers, request)
            }
            is InnsendingFeiletException -> {
                handleInnsendingFeiletException(ex, headers, request)
            }
            else -> {
                val status = HttpStatus.INTERNAL_SERVER_ERROR
                handleExceptionInternal(ex, ApiError(status.value(), INTERNAL_MSG), headers, status, request)
            }
        }
    }

    private fun handleConstraintViolationException(
        ex: ConstraintViolationException,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        return handleExceptionInternal(ex, ApiError(HttpStatus.BAD_REQUEST.value(), BAD_REQUEST_MSG), headers, HttpStatus.BAD_REQUEST, request)
    }

    private fun handleForbiddenException(
        ex: ForbiddenException,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.FORBIDDEN
        return handleExceptionInternal(ex, ApiError(status.value(), ex.message!!), headers, status, request)
    }

    private fun handleIllegalArgumentException(
        ex: IllegalArgumentException,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.BAD_REQUEST
        return handleExceptionInternal(ex, ApiError(status.value(), BAD_REQUEST_MSG), headers, status, request)
    }

    private fun handleJwtTokenUnauthorizedException(
        ex: JwtTokenUnauthorizedException,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.UNAUTHORIZED
        return handleExceptionInternal(ex, ApiError(status.value(), UNAUTHORIZED_MSG), headers, status, request)
    }

    private fun handleFastlegeIkkeFunnetException(
        ex: FastlegeIkkeFunnet,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.NOT_FOUND
        return handleExceptionInternal(ex, ApiError(status.value(), ex.message!!), headers, status, request)
    }

    private fun handlePartnerinformasjonIkkeFunnetException(
        ex: PartnerinformasjonIkkeFunnet,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.NOT_FOUND
        return handleExceptionInternal(ex, ApiError(status.value(), ex.message!!), headers, status, request)
    }

    private fun handleInnsendingFeiletException(
        ex: InnsendingFeiletException,
        headers: HttpHeaders,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        val status = HttpStatus.INTERNAL_SERVER_ERROR
        return handleExceptionInternal(ex, ApiError(status.value(), ex.message!!), headers, status, request)
    }

    private fun handleExceptionInternal(
        ex: Exception,
        body: ApiError,
        headers: HttpHeaders,
        status: HttpStatus,
        request: WebRequest
    ): ResponseEntity<ApiError> {
        metrikk.tellHttpKall(status.value())
        if (HttpStatus.INTERNAL_SERVER_ERROR == status) {
            log.error("Uventet feil: {} : {}", ex.javaClass.toString(), ex.message, ex)
            request.setAttribute(WebUtils.ERROR_EXCEPTION_ATTRIBUTE, ex, WebRequest.SCOPE_REQUEST)
        }
        return ResponseEntity(body, headers, status)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ControllerExceptionHandler::class.java)

        private const val BAD_REQUEST_MSG = "Vi kunne ikke tolke inndataene"
        private const val INTERNAL_MSG = "Det skjedde en uventet feil"
        private const val UNAUTHORIZED_MSG = "Autorisasjonsfeil"
    }
}
